Display smesher rewards correctly

brusherru opened the pull request:
It closes #1514 

For tests, I've used the wallet with a view-only account with a lot of rewards (ask me in Slack or Discord if you need it) and set the same address as Smesher's coinbase.

brusherru committed:
fix: sum rewards per layer coming from stream API instead of replacing values

brusherru committed:
fix: collect smesher rewards in the same way as wallet accounts

monikasmolarek commented:
@brusherru please hold on with this PR, I'll have a devnet to test it this night

monikasmolarek commented:
@brusherru, this devnet wasn’t the best and the testing could not be successful, do you want me do set it up on the testnet (epoch length is 24h) or are there any reliable steps to make to test it correctly on all 3 operating systems?
Edit: I did set up everything, so if you wish to wait, then let's get back to this in a couple of days

brusherru commented:
@monikasmolarek I've tested it right on the MainNet using the wallet file I've sent to you.

To do the tests you need to:
1. drop the node-data (better back it up of course),
2. remove `accounts` directory inside `%AppData%` directory
3. set up smeshing (using account `...evse`) -- you can pause smeshing even without having created any PoS data, but it will make smesher screen retrieving and displaying rewards for the selected coinbase
5. sync from scratch while having this wallet opened

Then just observe how rewards are coming, and that it is the same ones as in the transaction log.

If you do the same on the `develop` (or released Smapp up to v1.2.5 inclusive) you will notice the difference with the current PR: 
- sometimes on the smesher screen will be fewer rewards than in the transaction logs,
- calculated rewards on the smesher screen may be fewer than balance

maparr commented on app/redux/smesher/selectors.ts:
```diff
@@ -29,6 +37,54 @@ export const getSmeshingOpts = (state: RootState): SmeshingOpts => {
+
+  const dailyAverage =
+    rewards.length > 2
+      ? (sums.total /
```
Have a question, because we are performing division here, getLayerTime or sums.total ,to prevent getting Infinity/NaN as a result in case of 1 / 0 or 0/0, may we add an if state to prevent this situation?

brusherru replied:
We have a `Reward` type here (which is validated for every object we got from API within `parseReward` function). So there is no chance to have `undefined` or `0` layer number.

maparr commented on app/redux/smesher/selectors.ts:
```diff
@@ -29,6 +37,54 @@ export const getSmeshingOpts = (state: RootState): SmeshingOpts => {
+      ? (sums.total /
+          (getLayerTime(rewards[rewards.length - 1].layer) -
+            getLayerTime(rewards[0].layer))) *
+        (24 * 1000 * 60 * 60)
```
We are using already this value, it is time to make a const  

https://github.com/spacemeshos/smapp/blob/047c4cdeb38d9508609ed0abbc908b76d4879153/app/screens/transactions/Transactions.tsx#L239

maparr commented on app/redux/smesher/selectors.ts:
```diff
@@ -29,6 +37,54 @@ export const getSmeshingOpts = (state: RootState): SmeshingOpts => {
+  const curEpoch = getEpoch(curLayer);
+  const rewards = getRewards(state);
+
+  const sums = rewards.reduce(
```
may we separate sums and last epoch calculation and add tests for these cases?

maparr commented on desktop/AccountState.ts:
```diff
@@ -130,6 +130,9 @@ export class AccountStateManager {
     Object.values((this.state[this.genesisID] as StateType).rewards);
 
+  getRewardByLayer = (layer: number) =>
+    (this.state[this.genesisID] as StateType)?.rewards?.[layer];
```
```suggestion
    (this.state[this.genesisID] as StateType).rewards?.[layer];
```

maparr commented on desktop/TransactionManager.ts:
```diff
@@ -432,22 +451,55 @@ class TransactionManager extends AbstractManager {
+  };
+
+  addReward = (accountId: HexString) => (reward: RewardHandlerArg) => {
+    try {
+      const parsedReward = this.parseReward(reward);
+      const prevReward = this.accountStates[accountId].getRewardByLayer(
+        parsedReward.layer
+      );
+      if (prevReward) {
+        // Sum the reward amount and update
+        const nextReward: Reward = {
+          ...prevReward,
+          amount: prevReward.amount + parsedReward.amount,
+          layerReward: prevReward.layerReward + parsedReward.layerReward,
+        };
+        this.storeReward(accountId, nextReward);
+      } else {
+        // Store reward as is
+        this.storeReward(accountId, parsedReward);
+      }
+    } catch (err) {
+      this.logger.error('addReward', err, { accountId, reward });
+    }
```
```suggestion
     try {
    const parsedReward = this.parseReward(reward);
    const prevReward = this.accountStates[accountId].getRewardByLayer(parsedReward.layer);
    
    // If there's a previous reward, update it; otherwise, store the new reward
    const nextReward = prevReward
      ? {
       ...prevReward, 
       amount: prevReward.amount + parsedReward.amount, 
       layerReward: prevReward.layerReward + parsedReward.layerReward
      }
      : parsedReward;

    this.storeReward(accountId, nextReward);
  } catch (err) {
    this.logger.error('addReward', err, { accountId, reward });
  }
```

maparr commented on app/screens/node/Node.tsx:
```diff
@@ -360,12 +355,8 @@ const Node = ({ history, location }: Props) => {
-    (state: RootState) => state.smesher.rewardsInfo
-  );
+  const rewards = useSelector(SmesherSelectors.getRewards);
+  const rewardsInfo = useSelector(SmesherSelectors.getRewardsInfo);
```
For this selector, because of calculations, we can consider using memoization. It might be better to separate the functions, cover them with tests, and then have the selectors return only raw data. Alternatively, you can write your own hook for reuse or use useMemo here to avoid unnecessary recalculations.

brusherru replied:
Yeah, it might help a bit with re-opening Smesher screen (if memoization will be done in a global scope, not in the component). But it won't help with the case when Node is actively syncing (a lot of new rewards are coming every second).
So I will add memoization here and then try to optimize updating the list of rewards (and avoid re-calculating from scratch every time)...

maparr reviewed (commented):
is it already not prerequests https://github.com/spacemeshos/api/issues/272 ? based on the https://github.com/spacemeshos/smapp/issues/1514 description

maparr commented:
still checking based on the test case, need to sync on this branch to catch up on details in the code and debug

monikasmolarek commented:
@maparr if you will be checking also manually, please pay attention to the overall app behaviour. I have an impression that this PR-based build is very laggy and slow, it hangs up very often, especially while trying to open the Smesher screen. I compared to the 1.2.4 and 1.2.5 and I'm really afraid that something is off with it. 

Also the above-mentioned discrepancies in total sums/ balance. Also in this PR - The Wallet screen balance does not match the smesher or transaction log, even though there were very few transactions and the gas fee should not consume this much. (`Smesher: 230126.8077sмн` vs `Wallet: 230091.84 sмн`)
But this might be just refreshing lag (another one)? I'm not sure, ~35 sмн difference seems quite big, but with such a slow GUI I didn’t even try to analyse the rewards history. grpc aligned with wallet screen afair

brusherru committed:
fixup! fix: sum rewards per layer coming from stream API instead of replacing values

Co-authored-by: Maksym Parfenov <maksim.parfenov@protonmail.ch>

brusherru committed:
refactor: extract DAY constant and move MINUTE to shared space

brusherru committed:
refactor: optimize and fix displaying rewards on smesher screen

brusherru committed:
refactor: get rid of stale Reward type

brusherru committed:
refactor: remove useless `dailyAverage` and cover rewards info calculation with tests

brusherru commented:
1. Regarding the big diverge between balance and rewards — there was a bug. Fixed. However, there is still some chance of having a diverge because of the mentioned prerequisite https://github.com/spacemeshos/api/issues/272 (we still don't have the possibility to filter out duplicates, while two API endpoints return different results). Logout/login may help to get more actual info.
2. Also, when you log out and then log in — it may display only new coming rewards for a while. But then it will sync older rewards. This behavior may be fixed only after having a fixed API (same issue mentioned above).
3. I've optimized how rewards are pushed from the Main process to Renderer + added memoization to selectors. Now this screen (and especially when you're switching to it) should not freeze/lag.
4. Added tests for `RewardsInfo` calculation
5. Removed useless `dailyAverage` calculation

maparr reviewed (approved):
thanks for the update, working smoothly.
